File: ruth-ai-backend/app/schemas/analytics.py
"""Pydantic schemas for analytics API endpoints.

From API Contract - Analytics APIs.
Aligned with F6 §6 Analytics Domain and analytics-design.md Section 8.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class AnalyticsNotImplementedResponse(BaseModel):
    """Response for unimplemented analytics endpoints."""

    message: str = Field(
        default="Analytics endpoints not yet implemented",
        description="Status message",
    )
    endpoint: str = Field(..., description="Requested endpoint")
    phase: str = Field(default="Phase 4", description="Implementation phase")


class TimeRange(BaseModel):
    """Time range for analytics queries."""

    from_: datetime = Field(..., alias="from", description="Start of time range")
    to: datetime = Field(..., description="End of time range")

    class Config:
        populate_by_name = True


class AnalyticsTotals(BaseModel):
    """Analytics totals breakdown.

    Aligned with F6 §6.1 AnalyticsTotals interface and analytics-design.md §8.1.
    """

    violations_total: int = Field(default=0, description="Total violation count")
    violations_open: int = Field(default=0, description="Open violations count")
    violations_reviewed: int = Field(default=0, description="Reviewed violations count")
    violations_dismissed: int = Field(
        default=0, description="Dismissed violations count"
    )
    violations_resolved: int = Field(default=0, description="Resolved violations count")
    cameras_active: int = Field(default=0, description="Active cameras count")
    cameras_total: int = Field(default=0, description="Total cameras count")


class AnalyticsComparison(BaseModel):
    """Comparison to previous period."""

    violations_total_change: int = Field(
        default=0, description="Change in total violations vs previous period"
    )
    violations_total_change_percent: float = Field(
        default=0.0, description="Percentage change vs previous period"
    )


class CameraBreakdown(BaseModel):
    """Per-camera violation breakdown."""

    camera_id: str = Field(..., description="Camera UUID")
    camera_name: str = Field(..., description="Camera name")
    violations_total: int = Field(default=0, description="Total violations")
    violations_open: int = Field(default=0, description="Open violations")
    violations_reviewed: int = Field(default=0, description="Reviewed violations")
    violations_dismissed: int = Field(default=0, description="Dismissed violations")
    violations_resolved: int = Field(default=0, description="Resolved violations")


class TypeBreakdown(BaseModel):
    """Violation type breakdown."""

    type: str = Field(..., description="Violation type identifier")
    type_display: str = Field(..., description="Human-readable type name")
    count: int = Field(..., description="Number of violations")
    percentage: float = Field(..., description="Percentage of total (0-100)")


class StatusBreakdown(BaseModel):
    """Violation status breakdown."""

    status: str = Field(..., description="Violation status")
    count: int = Field(..., description="Number of violations")
    percentage: float = Field(..., description="Percentage of total (0-100)")


class TimeSeriesBucket(BaseModel):
    """Time series data bucket."""

    bucket: datetime = Field(..., description="Time bucket timestamp")
    total: int = Field(default=0, description="Total violations in this bucket")
    by_type: dict[str, int] = Field(
        default_factory=dict, description="Breakdown by type"
    )


class AnalyticsSummaryResponse(BaseModel):
    """Response schema for GET /api/v1/analytics/summary.

    Aligned with F6 §6.1 AnalyticsSummaryResponse interface and analytics-design.md §8.1.
    """

    time_range: TimeRange = Field(..., description="Query time range")
    totals: AnalyticsTotals = Field(..., description="Aggregated counts")
    comparison: AnalyticsComparison | None = Field(
        None, description="Comparison to previous period"
    )
    by_camera: list[CameraBreakdown] = Field(
        default_factory=list, description="Per-camera breakdown"
    )
    by_type: list[TypeBreakdown] = Field(
        default_factory=list, description="By violation type"
    )
    by_status: list[StatusBreakdown] = Field(
        default_factory=list, description="By status"
    )
    time_series: list[TimeSeriesBucket] = Field(
        default_factory=list, description="Time series data"
    )
    generated_at: datetime = Field(..., description="When summary was computed")


# Violations Trends API (§8.2)


class ViolationTrendBucket(BaseModel):
    """Trend data bucket with status breakdown."""

    bucket: datetime = Field(..., description="Time bucket timestamp")
    total: int = Field(default=0, description="Total violations in this bucket")
    by_type: dict[str, int] = Field(
        default_factory=dict, description="Breakdown by type"
    )
    by_status: dict[str, int] = Field(
        default_factory=dict, description="Breakdown by status"
    )


class ViolationTrendsResponse(BaseModel):
    """Response for GET /api/v1/analytics/violations/trends."""

    time_range: TimeRange = Field(..., description="Query time range")
    granularity: Literal["minute", "hour", "day"] = Field(
        ..., description="Time bucket granularity"
    )
    data: list[ViolationTrendBucket] = Field(..., description="Trend data")
    generated_at: datetime = Field(..., description="When data was computed")


# Device Status API (§8.2)


class DeviceAnalytics(BaseModel):
    """Analytics for a single device."""

    camera_id: str = Field(..., description="Camera UUID")
    camera_name: str = Field(..., description="Camera name")
    violations_total: int = Field(default=0, description="Total violations")
    violations_by_status: dict[str, int] = Field(
        default_factory=dict, description="Violations by status"
    )
    violations_by_type: dict[str, int] = Field(
        default_factory=dict, description="Violations by type"
    )
    avg_confidence: float = Field(default=0.0, description="Average confidence score")
    last_violation_at: datetime | None = Field(
        None, description="Timestamp of last violation"
    )


class DeviceStatusSummary(BaseModel):
    """Overall device status summary."""

    total_violations: int = Field(default=0, description="Total violations")
    total_cameras: int = Field(default=0, description="Total cameras")
    active_cameras: int = Field(default=0, description="Active cameras")


class DeviceStatusResponse(BaseModel):
    """Response for GET /api/v1/analytics/devices/status."""

    time_range: TimeRange = Field(..., description="Query time range")
    devices: list[DeviceAnalytics] = Field(..., description="Per-device analytics")
    summary: DeviceStatusSummary = Field(..., description="Overall summary")
    generated_at: datetime = Field(..., description="When data was computed")


# Export API (§8.2)


class ExportScope(BaseModel):
    """Scope definition for exports."""

    all: bool = Field(default=True, description="Export all violations")
    camera_ids: list[str] = Field(default_factory=list, description="Filter by cameras")
    violation_types: list[str] = Field(
        default_factory=list, description="Filter by types"
    )
    statuses: list[str] = Field(default_factory=list, description="Filter by statuses")


class ExportOptions(BaseModel):
    """Export format-specific options."""

    include_headers: bool = Field(default=True, description="Include column headers")
    include_timestamps: bool = Field(
        default=True, description="Include ISO 8601 timestamps"
    )
    include_raw_confidence: bool = Field(
        default=False, description="Include raw confidence scores"
    )
    include_evidence_urls: bool = Field(
        default=False, description="Include evidence URLs"
    )
    include_bounding_boxes: bool = Field(
        default=False, description="Include bounding box coordinates"
    )


class ExportRequest(BaseModel):
    """Request for POST /api/v1/analytics/export."""

    format: Literal["csv", "xlsx", "pdf"] = Field(..., description="Export format")
    time_range: TimeRange = Field(..., description="Time range to export")
    scope: ExportScope = Field(
        default_factory=ExportScope, description="Data scope filters"
    )
    options: ExportOptions = Field(
        default_factory=ExportOptions, description="Format-specific options"
    )

File: frontend/src/types/analytics.ts
/**
 * Analytics Domain Types
 *
 * Aligned with analytics-design.md §8 and backend schemas.
 * Per F6 §6 Analytics Domain.
 */

export interface TimeRange {
  from: string; // ISO 8601
  to: string; // ISO 8601
}

export interface AnalyticsTotals {
  violations_total: number;
  violations_open: number;
  violations_reviewed: number;
  violations_dismissed: number;
  violations_resolved: number;
  cameras_active: number;
  cameras_total: number;
}

export interface AnalyticsComparison {
  violations_total_change: number;
  violations_total_change_percent: number;
}

export interface CameraBreakdown {
  camera_id: string;
  camera_name: string;
  violations_total: number;
  violations_open: number;
  violations_reviewed: number;
  violations_dismissed: number;
  violations_resolved: number;
}

export interface TypeBreakdown {
  type: string;
  type_display: string;
  count: number;
  percentage: number;
}

export interface StatusBreakdown {
  status: string;
  count: number;
  percentage: number;
}

export interface TimeSeriesBucket {
  bucket: string; // ISO 8601
  total: number;
  by_type: Record<string, number>;
}

export interface AnalyticsSummaryResponse {
  time_range: TimeRange;
  totals: AnalyticsTotals;
  comparison: AnalyticsComparison | null;
  by_camera: CameraBreakdown[];
  by_type: TypeBreakdown[];
  by_status: StatusBreakdown[];
  time_series: TimeSeriesBucket[];
  generated_at: string; // ISO 8601
}

// Violation Trends API

export interface ViolationTrendBucket {
  bucket: string; // ISO 8601
  total: number;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
}

export interface ViolationTrendsResponse {
  time_range: TimeRange;
  granularity: 'minute' | 'hour' | 'day';
  data: ViolationTrendBucket[];
  generated_at: string; // ISO 8601
}

// Device Status API

export interface DeviceAnalytics {
  camera_id: string;
  camera_name: string;
  violations_total: number;
  violations_by_status: Record<string, number>;
  violations_by_type: Record<string, number>;
  avg_confidence: number;
  last_violation_at: string | null; // ISO 8601
}

export interface DeviceStatusSummary {
  total_violations: number;
  total_cameras: number;
  active_cameras: number;
}

export interface DeviceStatusResponse {
  time_range: TimeRange;
  devices: DeviceAnalytics[];
  summary: DeviceStatusSummary;
  generated_at: string; // ISO 8601
}

// Export API

export type ExportFormat = 'csv' | 'xlsx' | 'pdf';

export interface ExportScope {
  all: boolean;
  camera_ids?: string[];
  violation_types?: string[];
  statuses?: string[];
}

export interface ExportOptions {
  include_headers: boolean;
  include_timestamps: boolean;
  include_raw_confidence: boolean;
  include_evidence_urls: boolean;
  include_bounding_boxes: boolean;
}

export interface ExportRequest {
  format: ExportFormat;
  time_range: TimeRange;
  scope: ExportScope;
  options: ExportOptions;
}

// Time range presets

export type TimeRangePreset = '24h' | '7d' | '30d' | 'custom';

export interface TimeRangeSelection {
  preset: TimeRangePreset;
  from?: Date;
  to?: Date;
}

// Chart data helpers

export interface ChartDataPoint {
  label: string;
  value: number;
}

export interface LineChartSeries {
  name: string;
  data: Array<{ x: Date; y: number }>;
  color?: string;
}
